// @medusajs/product/dist/services/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProductImageService = exports.ProductOptionService = exports.ProductTypeService = exports.ProductVariantService = exports.ProductTagService = exports.ProductModuleService = exports.ProductCollectionService = exports.ProductCategoryService = exports.ProductService = void 0;
var product_1 = require("./product");
Object.defineProperty(exports, "ProductService", { enumerable: true, get: function () { return __importDefault(product_1).default; } });
var product_category_1 = require("./product-category");
Object.defineProperty(exports, "ProductCategoryService", { enumerable: true, get: function () { return __importDefault(product_category_1).default; } });
var product_collection_1 = require("./product-collection");
Object.defineProperty(exports, "ProductCollectionService", { enumerable: true, get: function () { return __importDefault(product_collection_1).default; } });
var product_module_service_1 = require("./product-module-service");
Object.defineProperty(exports, "ProductModuleService", { enumerable: true, get: function () { return __importDefault(product_module_service_1).default; } });
var product_tag_1 = require("./product-tag");
Object.defineProperty(exports, "ProductTagService", { enumerable: true, get: function () { return __importDefault(product_tag_1).default; } });
var product_variant_1 = require("./product-variant");
Object.defineProperty(exports, "ProductVariantService", { enumerable: true, get: function () { return __importDefault(product_variant_1).default; } });
var product_type_1 = require("./product-type");
Object.defineProperty(exports, "ProductTypeService", { enumerable: true, get: function () { return __importDefault(product_type_1).default; } });
var product_option_1 = require("./product-option");
Object.defineProperty(exports, "ProductOptionService", { enumerable: true, get: function () { return __importDefault(product_option_1).default; } });
var product_image_1 = require("./product-image");
Object.defineProperty(exports, "ProductImageService", { enumerable: true, get: function () { return __importDefault(product_image_1).default; } });
