// @medusajs/pricing/dist/loaders/container.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const defaultRepositories = __importStar(require("../repositories"));
const defaultServices = __importStar(require("../services"));
const utils_1 = require("@medusajs/utils");
const awilix_1 = require("awilix");
exports.default = async ({ container, options, }) => {
    const customRepositories = options?.repositories;
    container.register({
        currencyService: (0, awilix_1.asClass)(defaultServices.CurrencyService).singleton(),
        moneyAmountService: (0, awilix_1.asClass)(defaultServices.MoneyAmountService).singleton(),
        priceSetService: (0, awilix_1.asClass)(defaultServices.PriceSetService).singleton(),
        ruleTypeService: (0, awilix_1.asClass)(defaultServices.RuleTypeService).singleton(),
        priceSetMoneyAmountRulesService: (0, awilix_1.asClass)(defaultServices.PriceSetMoneyAmountRulesService).singleton(),
        priceRuleService: (0, awilix_1.asClass)(defaultServices.PriceRuleService).singleton(),
        priceSetRuleTypeService: (0, awilix_1.asClass)(defaultServices.PriceSetRuleTypeService).singleton(),
        priceSetMoneyAmountService: (0, awilix_1.asClass)(defaultServices.PriceSetMoneyAmountService).singleton(),
        priceListService: (0, awilix_1.asClass)(defaultServices.PriceListService).singleton(),
        priceListRuleService: (0, awilix_1.asClass)(defaultServices.PriceListRuleService).singleton(),
        priceListRuleValueService: (0, awilix_1.asClass)(defaultServices.PriceListRuleValueService).singleton(),
    });
    if (customRepositories) {
        (0, utils_1.loadCustomRepositories)({
            defaultRepositories,
            customRepositories,
            container,
        });
    }
    else {
        loadDefaultRepositories({ container });
    }
};
function loadDefaultRepositories({ container }) {
    container.register({
        baseRepository: (0, awilix_1.asClass)(defaultRepositories.BaseRepository).singleton(),
        pricingRepository: (0, awilix_1.asClass)(defaultRepositories.PricingRepository).singleton(),
        currencyRepository: (0, awilix_1.asClass)(defaultRepositories.CurrencyRepository).singleton(),
        moneyAmountRepository: (0, awilix_1.asClass)(defaultRepositories.MoneyAmountRepository).singleton(),
        priceSetRepository: (0, awilix_1.asClass)(defaultRepositories.PriceSetRepository).singleton(),
        ruleTypeRepository: (0, awilix_1.asClass)(defaultRepositories.RuleTypeRepository).singleton(),
        priceSetMoneyAmountRulesRepository: (0, awilix_1.asClass)(defaultRepositories.PriceSetMoneyAmountRulesRepository).singleton(),
        priceRuleRepository: (0, awilix_1.asClass)(defaultRepositories.PriceRuleRepository).singleton(),
        priceSetRuleTypeRepository: (0, awilix_1.asClass)(defaultRepositories.PriceSetRuleTypeRepository).singleton(),
        priceSetMoneyAmountRepository: (0, awilix_1.asClass)(defaultRepositories.PriceSetMoneyAmountRepository).singleton(),
        priceListRepository: (0, awilix_1.asClass)(defaultRepositories.PriceListRepository).singleton(),
        priceListRuleRepository: (0, awilix_1.asClass)(defaultRepositories.PriceListRuleRepository).singleton(),
        priceListRuleValueRepository: (0, awilix_1.asClass)(defaultRepositories.PriceListRuleValueRepository).singleton(),
    });
}
