// @medusajs/pricing/dist/models/price-rule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@mikro-orm/core");
const utils_1 = require("@medusajs/utils");
const price_set_1 = __importDefault(require("./price-set"));
const price_set_money_amount_1 = __importDefault(require("./price-set-money-amount"));
const rule_type_1 = __importDefault(require("./rule-type"));
let PriceRule = class PriceRule {
    beforeCreate() {
        this.id = (0, utils_1.generateEntityId)(this.id, "prule");
    }
};
__decorate([
    (0, core_1.PrimaryKey)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceRule.prototype, "id", void 0);
__decorate([
    (0, core_1.ManyToOne)({
        entity: () => price_set_1.default,
        fieldName: "price_set_id",
        name: "price_rule_price_set_id_unique",
        onDelete: "cascade",
        index: "IDX_price_rule_price_set_id",
    }),
    __metadata("design:type", price_set_1.default)
], PriceRule.prototype, "price_set", void 0);
__decorate([
    (0, core_1.ManyToOne)({
        entity: () => rule_type_1.default,
        fieldName: "rule_type_id",
        name: "price_rule_rule_type_id_unique",
        index: "IDX_price_rule_rule_type_id",
    }),
    __metadata("design:type", rule_type_1.default)
], PriceRule.prototype, "rule_type", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "boolean", default: false }),
    __metadata("design:type", Boolean)
], PriceRule.prototype, "is_dynamic", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceRule.prototype, "value", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "integer", default: 0 }),
    __metadata("design:type", Number)
], PriceRule.prototype, "priority", void 0);
__decorate([
    (0, core_1.ManyToOne)({
        onDelete: "cascade",
        entity: () => price_set_money_amount_1.default,
        fieldName: "price_set_money_amount_id",
        name: "price_set_money_amount_id_unique",
        index: "IDX_price_rule_price_set_money_amount_id",
    }),
    __metadata("design:type", price_set_money_amount_1.default)
], PriceRule.prototype, "price_set_money_amount", void 0);
__decorate([
    (0, core_1.BeforeCreate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], PriceRule.prototype, "beforeCreate", null);
PriceRule = __decorate([
    (0, core_1.Entity)()
], PriceRule);
exports.default = PriceRule;
