// @medusajs/pricing/dist/models/price-list-rule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@mikro-orm/core");
const utils_1 = require("@medusajs/utils");
const price_list_1 = __importDefault(require("./price-list"));
const price_list_rule_value_1 = __importDefault(require("./price-list-rule-value"));
const rule_type_1 = __importDefault(require("./rule-type"));
let PriceListRule = class PriceListRule {
    constructor() {
        this.price_list_rule_values = new core_1.Collection(this);
    }
    beforeCreate() {
        this.id = (0, utils_1.generateEntityId)(this.id, "plrule");
    }
};
__decorate([
    (0, core_1.PrimaryKey)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceListRule.prototype, "id", void 0);
__decorate([
    (0, core_1.ManyToOne)({
        entity: () => rule_type_1.default,
        fieldName: "rule_type_id",
        name: "price_rule_rule_type_id_unique",
        index: "IDX_price_list_rule_rule_type_id",
    }),
    __metadata("design:type", rule_type_1.default)
], PriceListRule.prototype, "rule_type", void 0);
__decorate([
    (0, core_1.OneToMany)(() => price_list_rule_value_1.default, (plrv) => plrv.price_list_rule, {
        cascade: [core_1.Cascade.REMOVE],
    }),
    __metadata("design:type", Object)
], PriceListRule.prototype, "price_list_rule_values", void 0);
__decorate([
    (0, core_1.ManyToOne)({
        entity: () => price_list_1.default,
        fieldName: "price_list_id",
        name: "price_rule_price_list_id",
        index: "IDX_price_list_rule_price_list_id",
    }),
    __metadata("design:type", price_list_1.default)
], PriceListRule.prototype, "price_list", void 0);
__decorate([
    (0, core_1.BeforeCreate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], PriceListRule.prototype, "beforeCreate", null);
PriceListRule = __decorate([
    (0, core_1.Entity)(),
    (0, core_1.Unique)({
        name: "IDX_price_list_rule_rule_type_id_price_list_id_unique",
        properties: ["price_list", "rule_type"],
    })
], PriceListRule);
exports.default = PriceListRule;
