// @medusajs/pricing/dist/models/price-set-money-amount.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const money_amount_1 = __importDefault(require("./money-amount"));
const price_list_1 = __importDefault(require("./price-list"));
const price_rule_1 = __importDefault(require("./price-rule"));
const price_set_1 = __importDefault(require("./price-set"));
const price_set_money_amount_rules_1 = __importDefault(require("./price-set-money-amount-rules"));
let PriceSetMoneyAmount = class PriceSetMoneyAmount {
    onCreate() {
        this.id = (0, utils_1.generateEntityId)(this.id, "psma");
    }
    constructor(money_amount, price_set) {
        this.price_rules = new core_1.Collection(this);
        this.price_set_money_amount_rules = new core_1.Collection(this);
        this.money_amount = money_amount;
        this.price_set = price_set;
    }
};
__decorate([
    (0, core_1.PrimaryKey)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceSetMoneyAmount.prototype, "id", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceSetMoneyAmount.prototype, "title", void 0);
__decorate([
    (0, core_1.ManyToOne)(() => price_set_1.default, {
        onDelete: "cascade",
        index: "IDX_price_set_money_amount_price_set_id",
    }),
    __metadata("design:type", price_set_1.default)
], PriceSetMoneyAmount.prototype, "price_set", void 0);
__decorate([
    (0, core_1.OneToOne)(() => money_amount_1.default, {
        onDelete: "cascade",
        index: "IDX_price_set_money_amount_money_amount_id",
    }),
    __metadata("design:type", money_amount_1.default)
], PriceSetMoneyAmount.prototype, "money_amount", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "integer", default: 0 }),
    __metadata("design:type", Number)
], PriceSetMoneyAmount.prototype, "number_rules", void 0);
__decorate([
    (0, core_1.OneToMany)({
        entity: () => price_rule_1.default,
        mappedBy: (pr) => pr.price_set_money_amount,
    }),
    __metadata("design:type", Object)
], PriceSetMoneyAmount.prototype, "price_rules", void 0);
__decorate([
    (0, core_1.OneToMany)({
        entity: () => price_set_money_amount_rules_1.default,
        mappedBy: (psmar) => psmar.price_set_money_amount,
    }),
    __metadata("design:type", Object)
], PriceSetMoneyAmount.prototype, "price_set_money_amount_rules", void 0);
__decorate([
    (0, core_1.ManyToOne)(() => price_list_1.default, {
        index: "IDX_price_rule_price_list_id",
        onDelete: "cascade",
        nullable: true,
    }),
    __metadata("design:type", price_list_1.default)
], PriceSetMoneyAmount.prototype, "price_list", void 0);
__decorate([
    (0, core_1.BeforeCreate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], PriceSetMoneyAmount.prototype, "onCreate", null);
PriceSetMoneyAmount = __decorate([
    (0, core_1.Entity)(),
    __metadata("design:paramtypes", [money_amount_1.default, price_set_1.default])
], PriceSetMoneyAmount);
exports.default = PriceSetMoneyAmount;
