// @medusajs/pricing/dist/models/price-list.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const utils_2 = require("@medusajs/utils");
const price_list_rule_1 = __importDefault(require("./price-list-rule"));
const price_set_money_amount_1 = __importDefault(require("./price-set-money-amount"));
const rule_type_1 = __importDefault(require("./rule-type"));
let PriceList = class PriceList {
    constructor() {
        this.price_set_money_amounts = new core_1.Collection(this);
        this.price_list_rules = new core_1.Collection(this);
        this.rule_types = new core_1.Collection(this);
    }
    onCreate() {
        this.id = (0, utils_1.generateEntityId)(this.id, "plist");
    }
};
__decorate([
    (0, core_1.PrimaryKey)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceList.prototype, "id", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceList.prototype, "title", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceList.prototype, "description", void 0);
__decorate([
    (0, core_1.Enum)({ items: () => utils_2.PriceListStatus, default: utils_2.PriceListStatus.DRAFT }),
    __metadata("design:type", String)
], PriceList.prototype, "status", void 0);
__decorate([
    (0, core_1.Enum)({ items: () => utils_2.PriceListType, default: utils_2.PriceListType.SALE }),
    __metadata("design:type", String)
], PriceList.prototype, "type", void 0);
__decorate([
    (0, core_1.Property)({
        columnType: "timestamptz",
        nullable: true,
    }),
    __metadata("design:type", Object)
], PriceList.prototype, "starts_at", void 0);
__decorate([
    (0, core_1.Property)({
        columnType: "timestamptz",
        nullable: true,
    }),
    __metadata("design:type", Object)
], PriceList.prototype, "ends_at", void 0);
__decorate([
    (0, core_1.OneToMany)(() => price_set_money_amount_1.default, (psma) => psma.price_list, {
        cascade: [core_1.Cascade.REMOVE],
    }),
    __metadata("design:type", Object)
], PriceList.prototype, "price_set_money_amounts", void 0);
__decorate([
    (0, core_1.OneToMany)(() => price_list_rule_1.default, (pr) => pr.price_list, {
        cascade: [core_1.Cascade.REMOVE],
    }),
    __metadata("design:type", Object)
], PriceList.prototype, "price_list_rules", void 0);
__decorate([
    (0, core_1.ManyToMany)({
        entity: () => rule_type_1.default,
        pivotEntity: () => price_list_rule_1.default,
        cascade: [core_1.Cascade.REMOVE],
    }),
    __metadata("design:type", Object)
], PriceList.prototype, "rule_types", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "integer", default: 0 }),
    __metadata("design:type", Number)
], PriceList.prototype, "number_rules", void 0);
__decorate([
    (0, core_1.Property)({
        onCreate: () => new Date(),
        columnType: "timestamptz",
        defaultRaw: "now()",
    }),
    __metadata("design:type", Date)
], PriceList.prototype, "created_at", void 0);
__decorate([
    (0, core_1.Property)({
        onCreate: () => new Date(),
        onUpdate: () => new Date(),
        columnType: "timestamptz",
        defaultRaw: "now()",
    }),
    __metadata("design:type", Date)
], PriceList.prototype, "updated_at", void 0);
__decorate([
    (0, core_1.Index)({ name: "IDX_price_list_deleted_at" }),
    (0, core_1.Property)({ columnType: "timestamptz", nullable: true }),
    __metadata("design:type", Date)
], PriceList.prototype, "deleted_at", void 0);
__decorate([
    (0, core_1.BeforeCreate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], PriceList.prototype, "onCreate", null);
PriceList = __decorate([
    (0, core_1.Entity)()
], PriceList);
exports.default = PriceList;
