// @medusajs/pricing/dist/models/price-list-rule-value.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@mikro-orm/core");
const price_list_rule_1 = __importDefault(require("./price-list-rule"));
const utils_1 = require("@medusajs/utils");
let PriceListRuleValue = class PriceListRuleValue {
    onCreate() {
        this.id = (0, utils_1.generateEntityId)(this.id, "plrv");
    }
};
__decorate([
    (0, core_1.PrimaryKey)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceListRuleValue.prototype, "id", void 0);
__decorate([
    (0, core_1.ManyToOne)(() => price_list_rule_1.default, {
        onDelete: "cascade",
        fieldName: 'price_list_rule_id',
        index: "IDX_price_list_rule_price_list_rule_value_id",
    }),
    __metadata("design:type", price_list_rule_1.default)
], PriceListRuleValue.prototype, "price_list_rule", void 0);
__decorate([
    (0, core_1.Property)({ columnType: "text" }),
    __metadata("design:type", String)
], PriceListRuleValue.prototype, "value", void 0);
__decorate([
    (0, core_1.BeforeCreate)(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], PriceListRuleValue.prototype, "onCreate", null);
PriceListRuleValue = __decorate([
    (0, core_1.Entity)()
], PriceListRuleValue);
exports.default = PriceListRuleValue;
