// @medusajs/pricing/dist/repositories/currency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrencyRepository = void 0;
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const _models_1 = require("../models");
class CurrencyRepository extends utils_1.DALUtils.MikroOrmBaseRepository {
    constructor({ manager }) {
        // @ts-ignore
        // eslint-disable-next-line prefer-rest-params
        super(...arguments);
        this.manager_ = manager;
    }
    async find(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.find(_models_1.Currency, findOptions_.where, findOptions_.options);
    }
    async findAndCount(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.findAndCount(_models_1.Currency, findOptions_.where, findOptions_.options);
    }
    async delete(codes, context = {}) {
        const manager = this.getActiveManager(context);
        await manager.nativeDelete(_models_1.Currency, { code: { $in: codes } }, {});
    }
    async create(data, context = {}) {
        const manager = this.getActiveManager(context);
        const currencies = data.map((currencyData) => {
            return manager.create(_models_1.Currency, currencyData);
        });
        manager.persist(currencies);
        return currencies;
    }
    async update(data, context = {}) {
        const manager = this.getActiveManager(context);
        const currencyCodes = data.map((currencyData) => currencyData.code);
        const existingCurrencies = await this.find({
            where: {
                code: {
                    $in: currencyCodes,
                },
            },
        }, context);
        const existingCurrencyMap = new Map(existingCurrencies.map((currency) => [
            currency.code,
            currency,
        ]));
        const currencies = data.map((currencyData) => {
            const existingCurrency = existingCurrencyMap.get(currencyData.code);
            if (!existingCurrency) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `Currency with code "${currencyData.code}" not found`);
            }
            return manager.assign(existingCurrency, currencyData);
        });
        manager.persist(currencies);
        return currencies;
    }
}
exports.CurrencyRepository = CurrencyRepository;
