// @medusajs/pricing/dist/repositories/money-amount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoneyAmountRepository = void 0;
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const _models_1 = require("../models");
class MoneyAmountRepository extends utils_1.DALUtils.MikroOrmBaseRepository {
    constructor({ manager }) {
        // @ts-ignore
        // eslint-disable-next-line prefer-rest-params
        super(...arguments);
        this.manager_ = manager;
    }
    async find(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.find(_models_1.MoneyAmount, findOptions_.where, findOptions_.options);
    }
    async findAndCount(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.findAndCount(_models_1.MoneyAmount, findOptions_.where, findOptions_.options);
    }
    async delete(ids, context = {}) {
        const manager = this.getActiveManager(context);
        await manager.nativeDelete(_models_1.MoneyAmount, { id: { $in: ids } }, {});
    }
    async create(data, context = {}) {
        const manager = this.getActiveManager(context);
        const moneyAmounts = data.map((moneyAmountData) => {
            return manager.create(_models_1.MoneyAmount, moneyAmountData);
        });
        manager.persist(moneyAmounts);
        return moneyAmounts;
    }
    async update(data, context = {}) {
        const manager = this.getActiveManager(context);
        const moneyAmountIds = data.map((moneyAmountData) => moneyAmountData.id);
        const existingMoneyAmounts = await this.find({
            where: {
                id: {
                    $in: moneyAmountIds,
                },
            },
        }, context);
        const existingMoneyAmountMap = new Map(existingMoneyAmounts.map((moneyAmount) => [
            moneyAmount.id,
            moneyAmount,
        ]));
        const moneyAmounts = data.map((moneyAmountData) => {
            const existingMoneyAmount = existingMoneyAmountMap.get(moneyAmountData.id);
            if (!existingMoneyAmount) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `MoneyAmount with id "${moneyAmountData.id}" not found`);
            }
            return manager.assign(existingMoneyAmount, moneyAmountData);
        });
        manager.persist(moneyAmounts);
        return moneyAmounts;
    }
}
exports.MoneyAmountRepository = MoneyAmountRepository;
