// @medusajs/pricing/dist/repositories/price-set-money-amount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PriceSetMoneyAmountRepository = void 0;
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const _models_1 = require("../models");
class PriceSetMoneyAmountRepository extends utils_1.DALUtils.MikroOrmBaseRepository {
    constructor({ manager }) {
        // @ts-ignore
        // eslint-disable-next-line prefer-rest-params
        super(...arguments);
        this.manager_ = manager;
    }
    async find(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.find(_models_1.PriceSetMoneyAmount, findOptions_.where, findOptions_.options);
    }
    async findAndCount(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.findAndCount(_models_1.PriceSetMoneyAmount, findOptions_.where, findOptions_.options);
    }
    async delete(ids, context = {}) {
        const manager = this.getActiveManager(context);
        await manager.nativeDelete(_models_1.PriceSetMoneyAmount, { id: { $in: ids } }, {});
    }
    async create(data, context = {}) {
        const manager = this.getActiveManager(context);
        const psma = data.map((psmaData) => {
            return manager.create(_models_1.PriceSetMoneyAmount, psmaData);
        });
        manager.persist(psma);
        return psma;
    }
    async update(data, context = {}) {
        const manager = this.getActiveManager(context);
        const ids = data.map((psmaData) => psmaData.id);
        const existingPriceSetMoneyAmounts = await this.find({
            where: {
                id: {
                    $in: ids,
                },
            },
        }, context);
        const existingPSMAMap = new Map(existingPriceSetMoneyAmounts.map((psma) => [psma.id, psma]));
        const priceSetMoneyAmounts = data.map((psmaData) => {
            const existingPSMA = existingPSMAMap.get(psmaData.id);
            if (!existingPSMA) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `PriceSetMoneyAmount with id "${psmaData.id}" not found`);
            }
            return manager.assign(existingPSMA, psmaData);
        });
        manager.persist(priceSetMoneyAmounts);
        return priceSetMoneyAmounts;
    }
}
exports.PriceSetMoneyAmountRepository = PriceSetMoneyAmountRepository;
