// @medusajs/pricing/dist/repositories/rule-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuleTypeRepository = void 0;
const utils_1 = require("@medusajs/utils");
const core_1 = require("@mikro-orm/core");
const _models_1 = require("../models");
class RuleTypeRepository extends utils_1.DALUtils.MikroOrmBaseRepository {
    constructor({ manager }) {
        // @ts-ignore
        // eslint-disable-next-line prefer-rest-params
        super(...arguments);
        this.manager_ = manager;
    }
    async find(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.find(_models_1.RuleType, findOptions_.where, findOptions_.options);
    }
    async findAndCount(findOptions = { where: {} }, context = {}) {
        const manager = this.getActiveManager(context);
        const findOptions_ = { ...findOptions };
        findOptions_.options ?? (findOptions_.options = {});
        Object.assign(findOptions_.options, {
            strategy: core_1.LoadStrategy.SELECT_IN,
        });
        return await manager.findAndCount(_models_1.RuleType, findOptions_.where, findOptions_.options);
    }
    async delete(ids, context = {}) {
        const manager = this.getActiveManager(context);
        await manager.nativeDelete(_models_1.RuleType, { id: { $in: ids } }, {});
    }
    async create(data, context = {}) {
        (0, utils_1.validateRuleAttributes)(data.map((d) => d.rule_attribute));
        const manager = this.getActiveManager(context);
        const ruleTypes = data.map((ruleTypeData) => {
            return manager.create(_models_1.RuleType, ruleTypeData);
        });
        manager.persist(ruleTypes);
        return ruleTypes;
    }
    async update(data, context = {}) {
        (0, utils_1.validateRuleAttributes)(data.map((d) => d.rule_attribute));
        const manager = this.getActiveManager(context);
        const ruleTypeIds = data.map((ruleType) => ruleType.id);
        const existingRuleTypes = await this.find({
            where: {
                id: {
                    $in: ruleTypeIds,
                },
            },
        }, context);
        const existingRuleTypesMap = new Map(existingRuleTypes.map((ruleType) => [
            ruleType.id,
            ruleType,
        ]));
        const ruleTypes = data.map((ruleTypeData) => {
            const existingRuleType = existingRuleTypesMap.get(ruleTypeData.id);
            if (!existingRuleType) {
                throw new utils_1.MedusaError(utils_1.MedusaError.Types.NOT_FOUND, `RuleType with id "${ruleTypeData.id}" not found`);
            }
            return manager.assign(existingRuleType, ruleTypeData);
        });
        manager.persist(ruleTypes);
        return ruleTypes;
    }
}
exports.RuleTypeRepository = RuleTypeRepository;
