// @medusajs/pricing/dist/joiner-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.joinerConfig = exports.entityNameToLinkableKeysMap = exports.LinkableKeys = void 0;
const modules_sdk_1 = require("@medusajs/modules-sdk");
const _models_1 = require("./models");
exports.LinkableKeys = {
    money_amount_id: _models_1.MoneyAmount.name,
    currency_code: _models_1.Currency.name,
    price_set_id: _models_1.PriceSet.name,
};
const entityLinkableKeysMap = {};
Object.entries(exports.LinkableKeys).forEach(([key, value]) => {
    entityLinkableKeysMap[value] ?? (entityLinkableKeysMap[value] = []);
    entityLinkableKeysMap[value].push({
        mapTo: key,
        valueFrom: key.split("_").pop(),
    });
});
exports.entityNameToLinkableKeysMap = entityLinkableKeysMap;
exports.joinerConfig = {
    serviceName: modules_sdk_1.Modules.PRICING,
    primaryKeys: ["id", "currency_code"],
    linkableKeys: exports.LinkableKeys,
    alias: [
        {
            name: "price_set",
        },
        {
            name: "price_sets",
        },
        {
            name: "money_amount",
            args: {
                methodSuffix: "MoneyAmounts",
            },
        },
        {
            name: "money_amounts",
            args: {
                methodSuffix: "MoneyAmounts",
            },
        },
        {
            name: "currency",
            args: {
                methodSuffix: "Currencies",
            },
        },
        {
            name: "currencies",
            args: {
                methodSuffix: "Currencies",
            },
        },
    ],
};
