// @medusajs/pricing/dist/services/price-set-rule-type.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@medusajs/types");
const utils_1 = require("@medusajs/utils");
const _models_1 = require("../models");
class PriceSetRuleTypeService {
    constructor({ priceSetRuleTypeRepository }) {
        this.priceSetRuleTypeRepository_ = priceSetRuleTypeRepository;
    }
    async retrieve(priceSetId, config = {}, sharedContext = {}) {
        return (await (0, utils_1.retrieveEntity)({
            id: priceSetId,
            entityName: _models_1.PriceSet.name,
            repository: this.priceSetRuleTypeRepository_,
            config,
            sharedContext,
        }));
    }
    async list(filters = {}, config = {}, sharedContext = {}) {
        return (await this.priceSetRuleTypeRepository_.find(this.buildQueryForList(filters, config), sharedContext));
    }
    async listAndCount(filters = {}, config = {}, sharedContext = {}) {
        return (await this.priceSetRuleTypeRepository_.findAndCount(this.buildQueryForList(filters, config), sharedContext));
    }
    buildQueryForList(filters = {}, config = {}) {
        const queryOptions = utils_1.ModulesSdkUtils.buildQuery(filters, config);
        if (filters.id) {
            queryOptions.where.id = { $in: filters.id };
        }
        return queryOptions;
    }
    async create(data, sharedContext = {}) {
        return (await this.priceSetRuleTypeRepository_.create(data, sharedContext));
    }
    async update(data, sharedContext = {}) {
        return (await this.priceSetRuleTypeRepository_.update(data, sharedContext));
    }
    async delete(ids, sharedContext = {}) {
        await this.priceSetRuleTypeRepository_.delete(ids, sharedContext);
    }
}
exports.default = PriceSetRuleTypeService;
__decorate([
    (0, utils_1.InjectManager)("priceSetRuleTypeRepository_"),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, Object, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "retrieve", null);
__decorate([
    (0, utils_1.InjectManager)("priceSetRuleTypeRepository_"),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "list", null);
__decorate([
    (0, utils_1.InjectManager)("priceSetRuleTypeRepository_"),
    __param(2, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Object, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "listAndCount", null);
__decorate([
    (0, utils_1.InjectTransactionManager)("priceSetRuleTypeRepository_"),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "create", null);
__decorate([
    (0, utils_1.InjectTransactionManager)("priceSetRuleTypeRepository_"),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "update", null);
__decorate([
    (0, utils_1.InjectTransactionManager)("priceSetRuleTypeRepository_"),
    __param(1, (0, utils_1.MedusaContext)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Array, Object]),
    __metadata("design:returntype", Promise)
], PriceSetRuleTypeService.prototype, "delete", null);
